File: fast_wadler/Pombrio.hs
-- An attempt at further simplification relative to lazy_dequeue.hs.
-- Optimized with help of Jonathan Sailor

-- This is my attempt at a major simplification of
-- Pretty Printing with Lazy Dequeues by Olaf Chitil.
--
-- (DOI:10.1145/1053468.1053473)
--
-- The idea is to _measure_ each Doc at construction time, determining enough information to resolve
-- choices without any lookahead.

module Pombrio (MDoc, nil, (<|>), (<>), nest, text, line, group, flatten, pretty) where
-- TODO: temp
type MDoc = Doc

{------------------------------------------------------------------------------}
{- Measurements                                                               -}
{------------------------------------------------------------------------------}

-- Int is flatLen         -- the minimum width if rendered flat
-- Maybe Int is suffixLen -- the width until the earliest possible newline, if any
data Measure = Measure Int (Maybe Int)

maxWidth = 256

emptyM :: Measure
emptyM = Measure 0 Nothing

flattenM :: Measure -> Measure
flattenM (Measure f s) = Measure f Nothing

-- Combine the measure of a Doc `x` and the measure of a Doc `y` to obtain the measure of `x <> y`.
addM :: Measure -> Measure -> Measure
addM m@(Measure f (Just s)) _ | s >= maxWidth = m
addM (Measure f (Just s)) (Measure f' _) = Measure (f + f') (Just s)
addM (Measure f Nothing) (Measure f' s') = Measure (f + f') suff
  where suff = case s' of
          Nothing -> Nothing
          Just s' -> Just (f + s')
--  addM (Measure f s) (Measure f' s') = Measure (f + f') (addSuffix f s s')
--    where
--    addSuffix _ Nothing  Nothing   = Nothing
--    addSuffix _ (Just s) Nothing   = Just s
--    addSuffix f Nothing  (Just s') = Just (f + s')
--    addSuffix f (Just s) (Just s') = Just s
--    addSuffix f (Just s) (Just s') = Just (s `min` (f + s'))

-- The width until the earliest possible newline, or end of document.
suffixLen :: Measure -> Int
suffixLen (Measure f Nothing) = f
suffixLen (Measure f (Just s)) = s
--suffixLen (Measure f (Just s)) = f `min` s

{------------------------------------------------------------------------------}
{- Pretty Printing                                                            -}
{------------------------------------------------------------------------------}

data Doc =
    Nil
  | Text String
  | Line
  | Nest Int Doc Measure
  | Flatten Doc Measure
  | Concat Doc Doc Measure
  | Choice Doc Doc Measure

measure :: Doc -> Measure
measure Nil = emptyM
measure (Text s) = Measure (length s) Nothing
measure Line = Measure 1 (Just 0)
measure (Nest i x m) = m
measure (Flatten x m) = m
measure (Concat x y m) = m
measure (Choice x y m) = m

group x = flatten x <|> x

nil = Nil
text s = Text s
line = Line
nest i x = Nest i x (measure x)
flatten x = Flatten x (flattenM (measure x))
x <> y = Concat x y (addM (measure x) (measure y))
x <|> y = Choice x y (measure y)

pretty :: Int -> Doc -> String
pretty w d = concat $ pp w 0 [(Just 0, emptyM, d)]
  where
    pp :: Int -> Int -> [(Maybe Int, Measure, Doc)] -> [String]
    pp w p [] = []
    pp w p ((_, _, Nil) : xs)              = pp w p xs
    pp w p ((_, _, Text s) : xs)           = s : pp w (p + length s) xs
    pp w p ((Nothing, m, Nest j x _) : ys) = pp w p ((Nothing, m, x) : ys)
    pp w p ((Just i,  m, Nest j x _) : ys) = pp w p ((Just (i + j), m, x) : ys)
    pp w p ((i, m, Flatten x _) : ys) = pp w p ((Nothing, m, x) : ys)
    pp w p ((i, m, Concat x y _) : zs) =
      pp w p ((i, addM (measure y) m, x) : (i, m, y) : zs)
    pp w p ((i, m, Choice x y _) : zs) =
      let itFits = p + (suffixLen (addM (measure x) m)) <= w
      in pp w p ((i, m, if itFits then x else y) : zs)
    pp w p ((i, m, Line) : xs) =
      case i of
        Nothing -> " " : pp w (p + 1) xs
        Just i  -> "\n" : replicate i ' ' : pp w i xs

File: fast_wadler/strict/Pombrio.ml

(****************
 * Measurements *
 ****************)

type measure = {
  flat : int;           (* the len if rendered flat *)
  nonflat : int option; (* the len until the earliest possible newline, if any *)
}

let empty_m = { flat = 0; nonflat = None }
let flatten_m { flat; nonflat } = { flat = flat; nonflat = None }

let add_suff f s1 s2 = match (s1, s2) with
  | (None, None)     -> None
  | (None, Some(s2)) -> Some(f + s2)
  | (Some(s1), _)    -> Some(s1)

(* Combine the measure of a Doc `x` and the measure of a Doc `y` to obtain the
 * measure of `x ^^ y`.  *)
let add_m m1 m2 = {
  flat = m1.flat + m2.flat;
  nonflat = add_suff m1.flat m1.nonflat m2.nonflat;
}

(************************
 * Data Representations *
 ************************)

(* Designed to match Lindig's Strictly Pretty for fair comparison. *)

type doc =
  | DocNil
  | DocText of string
  | DocBreak of string
  | DocFlatten of doc * measure
  | DocNest of int * doc * measure
  | DocConcat of doc * doc * measure
  | DocChoice of doc * doc * measure

let measure d = match d with
  | DocNil             -> empty_m
  | DocText(s)         -> { flat = String.length s; nonflat = None }
  | DocBreak(s)        -> { flat = String.length s; nonflat = Some(0) }
  | DocFlatten(_, m)   -> m
  | DocNest(_, _, m)   -> m
  | DocConcat(_, _, m) -> m
  | DocChoice(_, _, m) -> m

let empty       = DocNil
let text s      = DocText(s)
let break       = DocBreak(" ")
let breakWith s = DocBreak(s)
let flatten d   = DocFlatten(d, flatten_m (measure d))
let nest i x    = DocNest(i, x, measure x)
let (^^) x y    = DocConcat(x, y, add_m (measure x) (measure y))
let (^?) x y    = DocChoice(x, y, measure y)
let group d     = flatten d ^? d

(* The width until the earliest possible newline, or end of document. *)
let suffix_len = function
  | { flat; nonflat = None } -> flat
  | { flat; nonflat = Some(s) } -> s

(*******************
 * Pretty Printing *
 *******************)

type mode = | Flat | Ind of int

let rec format w k = function
  | []                                 -> []
  | (i, m, DocNil)                :: z -> format w k z
  | (i, m, DocText(s))            :: z -> s :: format w (k + String.length s) z
  | (Flat, m, DocBreak(s))        :: z -> s :: format w (k + String.length s) z
  | (Ind(i), m, DocBreak(s))      :: z -> "\n" :: String.make i ' ' :: format w i z
  | (Flat, m, DocNest(j, x, _))   :: z -> format w k ((Flat, m, x) :: z)
  | (Ind(i), m, DocNest(j, x, _)) :: z -> format w k ((Ind(i+j), m, x) :: z)
  | (i, m, DocConcat(x, y, _))    :: z ->
    let mx = add_m (measure y) m in
    format w k ((i, mx, x) :: (i, m, y) :: z)
  | (i, m, DocFlatten(x, _))      :: z -> format w k ((Flat, m, x) :: z)
  | (i, m, DocChoice(x, y, _))    :: z ->
    if k + suffix_len (add_m (measure x) m) <= w
    then format w k ((i, m, x) :: z)
    else format w k ((i, m, y) :: z)

let pretty w doc = String.concat "" (format w 0 [Ind(0), measure doc, doc])

(*************
 * Test Case *
 *************)

let rec bigdoc n =
  if n == 0
  then text "<leaf/>"
  else group
    ((nest 2 (text "<branch>" ^^ break ^^ bigdoc (n-1) ^^ break ^^ bigdoc (n-1)))
      ^^ break ^^ text "</branch>");;

print_endline (pretty 80 (bigdoc 14));;
(* On laptop: 0.13s real time, 0.04s user time *)
